src/libAtomVM/intn.c

```diff
@@ -0,0 +1,631 @@
+            19683, 21952, 24389, 27000, 29791, 0, 1089, 1156, 1225, 1296
+        };
+    */
+
+    static const uint8_t pad[] = { 14, 9, 7, 6, 5, 5, 4, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3,
+        3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2 };
+
+    // let's keep space for abs(INT_MIN), that is bigger than INT_MAX
+    // and it must be supported, since we must allow converting to string INT_MIN as well
+    int tmp_buf_size = (256 / (sizeof(uint32_t) * 8)) + 1;
+    uint32_t tmp_buf1[tmp_buf_size];
+    uint32_t tmp_buf2[tmp_buf_size];
+
+    char *outbuf = malloc(257);
+    if (IS_NULL_PTR(outbuf)) {
+        return NULL;
+    }
+    char *end = outbuf + 256;
+    *end = '\0';
+
+    uint16_t *u;
+    size_t m;
+
+    bool negative_integer = is_negative(num, len);
+
+    if (negative_integer) {
+        size_t m_abs;
+        intn_abs(num, len, tmp_buf1, &m_abs);
+        m = m_abs;
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+        big_endian_in_place_swap_16(tmp_buf1, m);
+#endif
+    } else {
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+        memcpy(tmp_buf1, num, len * sizeof(uint32_t));
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+        uint16_t *dest_buf = (uint16_t *) tmp_buf1;
+        const uint16_t *num16 = (const uint16_t *) num;
+        for (size_t i = 0; i < len * 2; i += 2) {
+            dest_buf[i] = num16[i + 1];
+            dest_buf[i + 1] = num16[i];
+        }
+#endif
+        m = len;
+    }
+    u = (uint16_t *) tmp_buf1;
+
+    int m16 = count16(u, m * 2);
+
+    uint16_t *q = (uint16_t *) tmp_buf2;
+
+    do {
+        uint16_t r;
+
+        // divide in chunks that can be converted later
+        // using a bigger divisor like 10000 reduces the calls to this function
+        // so regular division on a smaller number can be used later
+        // example: 123456789 % 10000 = 6789, 123456789 / 10000 = 12345
+        if (UNLIKELY(divmnu16(q, &r, u, &bases[base - 2], m16, 1) != 0)) {
+            abort();
+        }
+
+        size_t intlen = intptr_write_to_ascii_buf(r, base, end);
+        end -= intlen;
+
+        m16 = count16(q, m16);
+
+        // add padding: `intptr_write_to_ascii_buf(7, 10, ptr)` will write just '7',
+        // but when dealing with base 10 we need 0007
+        // in order to handle numbers such as 110007 (note: 110007  / 10000 -> 11.0007, those
+        // digits cannot be discarded)
+        if (m16) {
+            int padsize = pad[base] - intlen;
+            end -= padsize;
+            for (int i = 0; i < padsize; i++) {
+                end[i] = '0';
+            }
+        }
+
+        // swap q (output) and u (input) buffers
+        uint16_t *swap_tmp = u;
+        u = q;
+        q = swap_tmp;
+    } while (m16 != 0);
+
+    if (negative_integer) {
+        end -= 1;
+        *end = '-';
+    }
+
+    size_t str_size = 257 - (end - outbuf);
+    memmove(outbuf, end, str_size);
+
+    *string_len = str_size - 1;
+    return realloc(outbuf, str_size);
```

Comment:
realloc result must be checked. If it returns NULL, outbuf is not freed.

Reply:
fixed.